### ISSUE TITLE
Random crash

### ISSUE THREAD

**Issue by Maeda1** [jaor/xmobar#310]
Hi there,

I'm using Xmobar with Xmonad tilling desktop manger, with two Xmobar bars.

Randomly, Xmobar crashes. I executed one from terminal, and when it occurs (not very often, but more often if computer's CPU is heavily used), here is what I get : 
```
X Error of failed request:  BadDrawable (invalid Pixmap or Window parameter)
  Major opcode of failed request:  139 (RENDER)
  Minor opcode of failed request:  4 (RenderCreatePicture)
  Resource id in failed request:  0x3a00001
  Serial number of failed request:  505915
  Current serial number in output stream:  505952
```
Some informations : 

1. Crashes occurs with the two bars, but more more often with second one started manually than the one started along with Xmonad.
2. Not very important, but when I start this secondary Xmobar, output is :
```
Fields missing from config defaulted: additionalFonts,border,borderColor,textOffset,iconOffset,allDesktops,overrideRedirect,pickBroadest,lowerOnStart
```
3. I'm using some pictures (may be related to 'RenderCreatePicture' error ?) on my bars.

My system : 
- 4.12.4-1-ARCH
- xmonad 0.13
- xmobar 0.24.5

[my xmobarrc2 file](http://lpaste.net/3445859674694352896)

**Comment by jaor** [comment]
Hi, and thanks for the detailed report.  I've never seen anything of
this sort, and i am using also xmonad (or sometimes sawfish) with two
xmobar instances.  I agree with your suspicion that the problem might be
related with images, although i don't see off hand how a higher CPU
usage might influence the problem.  I guess your icon files are on a
regular filesystem folder that is not mounted in any special way
(nothing exotic, like a network mount or something).  If so (and you
feel like it!) maybe we could try to confirm images are indeed the
problem by running your xmobar instances without icons?  Another very
silly thing one could try is using an icon path that contains only pure
ascii characters (something like "Images/icons" instead of
"Images/Icônes"): although i would be enormously surprised is that were
the cause of the problem, one never knows with these old X11 apis :)

**Comment by Maeda1** [comment]
Indeed, pictures are on local hard disk, ext4 filesystem, no exotic things here :)
Thanks for advices, I'll test them and will report here.

Stay tuned 👍

**Comment by Maeda1** [comment]
For testing, I commented out the line ~/Images/Icônes.
Then I started the secondary bar without any picture, and crashed several hours later. 

Here's the output : 
```
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
X Error of failed request:  BadDrawable (invalid Pixmap or Window parameter)
  Major opcode of failed request:  139 (RENDER)
  Minor opcode of failed request:  4 (RenderCreatePicture)
  Resource id in failed request:  0x3200001
  Serial number of failed request:  949956
  Current serial number in output stream:  949993
```
It doesn't seems to be related to pictures.

**Comment by andrewcchen** [comment]
I'm also having this issue. It seems to be related to changes in ghc 8.2.1 because the same code compiled with ghc 8.0.2 (stack lts-9.1) does not crash.

**Comment by octoploid** [comment]
See: https://ghc.haskell.org/trac/ghc/ticket/13707

**Comment by Maeda1** [comment]
Pertinent one.
I can't register to ghc to post and help, there is an internal server error :(
[ghc register page](https://ghc.haskell.org/trac/ghc/register)

**Comment by MP2E** [comment]
Just wanted to post that I'm getting the same issue on NixOS, here is my dmesg output

[12695.945204] xmobar[1185]: segfault at 12 ip 00007f957ea083e3 sp 00007ffe621969b8 error 4 in libHSghc-prim-0.5.1.0-ghc8.2.1.so[7f957e663000+3c6000]

**Comment by Haron-Prime** [comment]
I have the same problems after the update
I tried using XmonadLog, StdinReader and UnsafeStdinReader. Most often crash when using XmonadLog
Dmesg output:
[  138.348194] traps: xmobar[853] general protection ip:7f664f29db0c sp:7ffd00cd1b60 error:0 in libHSrts-ghc8.2.1.so[7f664f270000+5c000]

**Comment by Emantor** [comment]
Building xmobar with `--enable-debug-info=1` seems to unfortunately fix the segfaults.

**Comment by sugry** [comment]
I also have this issue.
Kernel 4.12.13-1-ARCH
xmonad 0.13-8
xmobar 0.24.5-6

**Comment by rafaqz** [comment]
I have the same issue since upgrading ghc to 8.2.1. It wasn't an issue on ghc 8.0 and that's the only thing that has changed in my xmonad setup. Xmobar 0.24.5, xmonad 0.13

**Comment by Rogach** [comment]
Had the same issue - segfaults, fixed by `--enable-debug-info=1`. However, crashes were not random - managed to reproduce them by dragging my xmonad floating windows over xmobar window, this resulted in xmobar crashing.

**Comment by perglr** [comment]
The same here: standard Arch Linux packages, i.e.
xmonad 0.13-9
xmonad-contrib 0.13-7
xmobar 0.24.5-6
"Random" crashes, dragging a window over xmobar crashes that for sure.

**Comment by yourealwaysbe** [comment]
Same here, also Arch Linux (xmobar 0.24.5-6, xmonad 0.13-9, xmonad-contrib 0.13-7, ghc 8.2.1-2, ghc-libs 8.2.1-2).  Compiling the git version (xmobar-git in the Arch AUR) with --enable-debug-info=1 also seems to fix the issue for me.

**Comment by ikyp** [comment]
I think is a *Arch* `ghc` package issue.

Because the *Arch* `ghc 8.2.1` even couldn't  compile simple program when i updated from `pacman` on Arch.

e.g.
```Haskell
plus :: Num a => a -> a -> a
plus x y = x + y

main :: IO ()
main = do
  print (plus 4.5 5.5)
``` 
show following error messge:

> Could not find module ‘Prelude’
>     There are files missing in the ‘base-4.10.0.0’ package,
>     try running 'ghc-pkg check'.
>     Use -v to see a list of the files searched for.

If i compile a complex program, more error message than above.

**Comment by Rogach** [comment]
@ikyp - Arch's ghc packaging feels screwed, right, but I don't think that it's the cause of the issue. In your case, manually running hooks in `/usr/share/haskell/register/` should do the trick.

**Comment by Haron-Prime** [comment]
Compiled with --enable-debug-info = 1 - no more crashes

**Comment by LeifW** [comment]
@ikyp I don't think that's related. As of GHC 8.0.2, Arch Linux GHC packages have switched to being dynamically linked by default - there are no static libs installed. Pass the `-dynamic` flag to GHC when compiling your program. See https://wiki.archlinux.org/index.php/haskell#Compiler

**Comment by ikyp** [comment]
@Rogach  **Thanks.**

@LeifW  `Big thanks.`  **I don't know these changed.** I used by compile & install from offical binary `GHC 8.2.1` version these days.

**Comment by andrewcchen** [comment]
I filed a ghc bug here: https://ghc.haskell.org/trac/ghc/ticket/14346

**Comment by ign1s-fatuus** [comment]
I don't know if this is the same issue, but I've been having issues with xmobar crashing randomly only after building them with requisite flags (with_alsa, xpm, &c.) and launching it via xmonad. Eventually it will certainly crash on either system I am running. I have tried to compile it with "--enable-debug-info=1" but it doesn't seem to make much a difference. The odd thing that I have found is that if I launch xmobar from the terminal, not listening to StdinReader, it acts in a rather stable manner. I have since compiled it via a number of different ghc builds and find that the random crashes will occur when called via xmonad at startup.

**Comment by jaor** [comment]
If the problem turns out to be *starting* xmobar within xmonad.hs to use
StdinReader, and easy workaround is to start xmobar by itself and use
XMonadLog instead of StdinReader to get the window titles/workspace
numbers, etc.  There's a simple example of how to use XMonadLog (and its
unsafe alternative) at the end of [this
section](https://archives.haskell.org/projects.haskell.org/xmobar/#other-plugins)
in the manual.

**Comment by Emantor** [comment]
I can't confirm that, I'm using XMonadLog with xmobar running in a systemd user unit.

The commentary on the ghc bug from @andrewcchen indicates that there is a real GHC Bug here, but I think this issue consists of a mixup of two different bugs.

**Comment by ign1s-fatuus** [comment]
Thanks for the reply and I'll surely look into using the StdinReader in an alternative way. It turns out that after a couple of hours it did crash as can be found in my edit: 

> --edit ... I seemed to have spoke too soon--xmobar crashed even when launched from the terminal. Failed with:
>
> X Error of failed request: BadDrawable (invalid Pixmap or Window parameter)
> Major opcode of failed request: 139 (RENDER)
> Minor opcode of failed request: 4 (RenderCreatePicture)
> Resource id in failed request: 0x3200001
> Serial number of failed request: 15320401
> Current serial number in output stream: 15320518

I have compiled and run my xmobar setup on two different distros (Arch and Ubuntu). My Arch build used ghc 8.2.1 I believe and my Ubuntu build is using ghc 7.10.3. While it seems that the latest version of ghc is the culprit as discussed above, I am surprised to see the same behavior in the one compiled with an earlier (and perhaps arguably more stable) version of ghc. 

I'll poke around, play with settings, and see if I can find a way to reliably induce the crash state.

**Comment by b0o** [comment]
I've been experiencing this bug as well. Seemingly random crashes quite often, every 5-30 minutes. I had never noticed it but I can confirm that dragging floating windows around rapidly will cause a crash, as @Rogach mentioned. 

Compiling with `--enable-debug-info=1` also solved my problem, but I figured I'd add some details here in case they're helpful:

```
$ uname -srmo 
Linux 4.13.7-1-ARCH x86_64 GNU/Linux

$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.2.1

$ cabal --version
cabal-install version 1.24.0.2
compiled using version 1.24.2.0 of the Cabal library 

$ xmobar --version
xmobar 0.24.5
```

[My xmobar.conf](https://github.com/b0o/xmonad-conf/blob/master/xmobar.conf)

[My xmobar init script (called from inside xmonad)](https://github.com/b0o/xmonad-conf/blob/master/initbars.sh)

[My xmonad log hook(s) and logic](https://github.com/b0o/xmonad-conf/blob/master/lib/Conf/Hooks/Log.hs)

**Comment by octoploid** [comment]
Fixed upstream by: http://git.haskell.org/ghc.git/commitdiff/404bf05ed3193e918875cd2f6c95ae0da5989be2

**Comment by MP2E** [comment]
Since updating to latest ghc-8.2 branch on git, I cannot reproduce this issue anymore

**Comment by Maeda1** [comment]
Solved for me too. I'll open another issue, as mouse clicks seems to do dumped core.
